src/tracker/inflights.rs

```diff
@@ -37,13 +42,38 @@ impl Inflights {
             start: 0,
             count: 0,
             cap,
+            incoming_cap: None,
+        }
+    }
+
+    /// Adjust inflight buffer capacity. Set it to `0` will disable the progress.
+    pub fn set_cap(&mut self, incoming_cap: usize) {
+        match self.cap.cmp(&incoming_cap) {
+            Ordering::Equal => self.incoming_cap = None,
+            Ordering::Less => {
+                self.buffer.reserve(incoming_cap - self.count);
+                self.cap = incoming_cap;
+                self.incoming_cap = None;
+            }
+            Ordering::Greater => {
+                if self.count == 0 {
+                    debug_assert!(self.incoming_cap.is_none());
+                    self.cap = incoming_cap;
+                    if self.buffer.capacity() > 0 {
+                        self.buffer = Vec::with_capacity(incoming_cap);
+                    }
+                } else {
+                    // TODO: is it better to shrink if count is less than incoming_cap?
+                    self.incoming_cap = Some(incoming_cap);
```

Comment:
How about enabling both of them, at least for future testing?

Reply:
In my opinion, we should shrink the buffer in this case, so we can enforce reallocating memory in case we are running out of space.

Reply:
It's updated. PTAL again, thank you!